[nn] Add remove_duplicate flag to named_buffers (#674)

jerryzh168 opened the pull request:
Summary:
X-link: https://github.com/pytorch/torchrec/pull/674

Pull Request resolved: https://github.com/pytorch/pytorch/pull/84984

this is to allow named_buffers to return the same buffer objects with different names multiple times, needed by internal use cases
ghstack-source-id: 168589597

Test Plan:
python test/test_nn.py -k test_buffers_and_named_buffers

Imported from OSS

Reviewed By: albanD

Differential Revision: D39493161

albanD commented:
What is the difference with the other PR?

jerryzh168 commented:
> What is the difference with the other PR?

I won't land the other PR. this PR is generated because:
* I imported the initial PR to fbcode
* I found that there are changes needed to other repos, torchrec etc.
* so I made modifications in fbcode
* I need to export from the fbcode diff, so I unlinked the other PR and exported this one

albanD commented on torch/nn/modules/module.py:
```diff
@@ -1765,6 +1766,7 @@ def named_buffers(self, prefix: str = '', recurse: bool = True) -> Iterator[Tupl
             recurse (bool): if True, then yields buffers of this module
                 and all submodules. Otherwise, yields only buffers that
                 are direct members of this module.
+            remove_duplicate: whether to remove the duplicated buffers in the result
```
nit: needs type and default value as: `remove_duplicate (bool, optional): whether to remove the duplicated buffers in the result. Defaults to True.`
The arguments above should be updates as well.

jerryzh168 replied:
since it requires updating not just oss apis (some internal apis are using these args as well), I'll put a TODO for the other arg for now

jerryzh168 replied:
OK, only found to occurrences actually, I'll fix that in this PR then

albanD reviewed (approved):
Sounds good, only minor nit.

jerryzh168 committed:
[nn] Add remove_duplicate flag to named_buffers (#85903)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/85903

X-link: https://github.com/pytorch/torchrec/pull/674

Pull Request resolved: https://github.com/pytorch/pytorch/pull/84984

this is to allow named_buffers to return the same buffer objects with different names multiple times, needed by internal use cases
ghstack-source-id: 168589597

Test Plan:
python test/test_nn.py -k test_buffers_and_named_buffers

Imported from OSS

Reviewed By: HDCharles, albanD

Differential Revision: D39493161

fbshipit-source-id: 539d379d0952d2e9cbb8d558db7d129cce5c7005
